### PR TITLE
Provide Position API to ad

### PR DESCRIPTION
**Pull request by zhouyx**
For #6873 
@alanorozco  I didn't make change  to `integration.js` When would `3p-use-ampcontext` be live?

Note: We once agreed to only send position if ad is in viewport. However to support inabox current logic, we need to send position even outside viewport. Not sure we want to do that. 

I only add the API to `<amp-ad>` not `<amp-iframe>` since there's no requirement for iframe case.
Need to document the API afterwards. 

cc @jasti

### CODE REVIEW

File: src/position-observer.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        this.subscriptionApi_.send(MessageType.POSITION, this.getPosition_());
+      });
+    }, MIN_EVENT_INTERVAL_IN_MS);
+
+    /** @private {?Function} */
+    this.unlistenOnDestroy_ = null;
+  }
+
+  /**
+   * Function to start listening to viewport event and observe element position.
```
Comment (alanorozco) [inline src/position-observer.js]: Nit: Change to "Starts listening to viewport events to observe the element's position."

File: src/position-observer.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+   * @private
+   */
+  startSendingPosition_() {
```
Comment (alanorozco) [inline src/position-observer.js]: The subject of the method should be the class instance, so "startSendingPosition" is a little confusing since the verb's subject is the viewport.

How about `startListening_`?

File: src/position-observer.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    // TODO: register element to positionObserver once layer manager supports
+    const unlistenViewportScroll = this.viewport_.onScroll(this.fire);
+    const unlistenViewportChange = this.viewport_.onChange(this.fire);
+    this.unlistenOnDestroy_ = () => {
```
Comment (alanorozco) [inline src/position-observer.js]: Nit:

How about having a class-level array of unlisteners `this.unlistenOnDestroy_ = []`, then changing this method to:

```
this.unlistenOnDestroy_.push(this.viewport_.onScroll(this.fire));
this.unlistenOnDestroy_.push(this.viewport_.onChange(this.fire));
```

...and then pull from the array in the `destroy()` method to execute each unlistener?

File: src/position-observer.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    /** @private {?SubscriptionApi} */
+    this.subscriptionApi_ = new SubscriptionApi(
+        iframe, MessageType.SEND_POSITIONS, opt_is3p || false, () => {
+          // Sending one immediately
+          this.fire();
+          this.startSendingPosition_();
+        });
+
+    /** @const {function()} */
+    this.fire = rateLimit(this.baseElement_.win, () => {
```
Comment (alanorozco) [inline src/position-observer.js]: Should `this.fire` be private?

File: src/position-observer.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+export class PositionObserverApi {
+  /**
+   * @param {!AMP.BaseElement} baseElement
+   * @param {!Element} iframe
+   * @param {boolean=} opt_is3p
+   */
+  constructor(baseElement, iframe, opt_is3p) {
+    /** @private @const {!AMP.BaseElement} */
+    this.baseElement_ = baseElement;
```
Comment (alanorozco) [inline src/position-observer.js]: Why does this have to be an AMP element necessarily?

File: test/functional/test-position-observer.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    sandbox.stub(baseElement, 'getLayoutBox', () => {
+      return layoutRectLtwh(5, 6, 7, 8);
+    });
+
+    const iframe = env.win.document.createElement('iframe');
+    iframe.src = 'no-use';
+    posObserver = new PositionObserverApi(baseElement, iframe);
+  });
+
+  it('should send one immediately wit correct value', () => {
```
Comment (alanorozco) [inline test/functional/test-position-observer.js]: "with correct value"

File: src/position-observer.js (hunk 3 above)
Comment (alanorozco) [inline src/position-observer.js]: Actually, I realized I misunderstood the concepts. So maybe not naming it `PositionObserver` at all :)

It's a little confusing to have a `PositionObserver` (`ads/inabox/position-observer.js`) and a `PositionObserverApi` since the purpose of both seems like it's the same, but it's not.

I suggest changing the name to reflect that this will be an *emitter* rather than a *listener*.

File: src/position-observer.js [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    }
+    if (this.subscriptionApi_) {
+      this.subscriptionApi_.destroy();
+      this.subscriptionApi_ = null;
+    }
+  }
+
+  /**
+   * Get the position info.
+   * Note: need to call in vsync
```
Comment (alanorozco) [inline src/position-observer.js]: I think if it needs to be wrapped in a measurer anyway, I suggest two options:

a.

Extract a method that sends the position message and call that from `this.fire`.  Kinda like:

```javascript
this.fire = rateLimit(this.baseElement_.win, () => {
  this.sendPositionMessage_();
}, MIN_EVENT_INTERVAL_IN_MS);

//...

sendPositionMessage_() {
  this.baseElement_.getVsync().measure(() => {
    this.subscriptionApi_.send(MessageType.POSITION, {
      viewport: this.viewport_.getRect(),
      target: this.baseElement_.getLayoutBox(),
    });
  });
}
```

or 

b.

Ensure that `getPosition_()` is always executed in a vsync measure context by returning a promise:

```javascript
this.fire = rateLimit(this.baseElement_.win, () => {
  this.getPosition_().then(position => {
    this.subscriptionApi_.send(MessageType.POSITION, position);
  });
}, MIN_EVENT_INTERVAL_IN_MS);

// ...

getPosition_() {
  return new Promise(resolve => {
    this.baseElement_.getVsync().measure(() => {
       resolve({
         viewport: this.viewport_.getRect(),
         target: this.baseElement_.getLayoutBox(),
      });
    });
  });
}
```

or...

c.

Do both!

```javascript
this.fire = rateLimit(this.baseElement_.win, () => {
  this.sendPositionMessage_();
}, MIN_EVENT_INTERVAL_IN_MS);

//...

sendPositionMessage_() {
  this.getPosition_().then(position => {
    this.subscriptionApi_.send(MessageType.POSITION, position);
  });
}

getPosition_() {
  return new Promise(resolve => {
    this.baseElement_.getVsync().measure(() => {
       resolve({
         viewport: this.viewport_.getRect(),
         target: this.baseElement_.getLayoutBox(),
      });
    });
  });
}
```

File: src/position-observer.js (hunk 2 above)
Comment (alanorozco) [inline src/position-observer.js]: Actually, I realized I misunderstood the concepts. So maybe not naming it `PositionObserver` at all :)

It's a little confusing to have a `PositionObserver` (`ads/inabox/position-observer.js`) and a `PositionObserverApi` since the purpose of both seems like it's the same, but it's not.

I suggest changing the name to reflect that this will be an *emitter* rather than a *listener*.

File: test/functional/test-position-observer.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+      return {
+        measure: function(callback) {
+          callback();
+        },
+      };
+    });
+    sandbox.stub(baseElement, 'getViewport', () => {
+      return {
+        getRect: () => {
+          console.log("asdfsdf");
```
Comment (zhouyx) [inline test/functional/test-position-observer.js]: console.log("oops")

File: src/position-observer.js [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+/**
+ * @typedef {{
+ *   viewport: !LayoutRectDef,
+ *   target: !LayoutRectDef
+ * }}
+ */
+let PositionEntryDef;
+
+export class PositionObserverApi {
```
Comment (zhouyx) [inline src/position-observer.js]: As you also pointed out `PositionObserver` is not a good name. I name it `PositionObserverApi` because we have a `IntersectionObserverApi` class, and I just name it blindly : )

File: src/position-observer.js (hunk 5 above)
Comment (zhouyx) [inline src/position-observer.js]: Ideally this should be any element. However that would require me to call `getBoundingClientRect()` which is expensive I think. Given that it is an AMP element, I can get viewport rect and element layoutBox with no extra cost. Also AMP element can deal with current use cases, I would like to make it AMP element unless there're further requirements in the future.

File: src/position-observer.js (hunk 4 above)
Comment (zhouyx) [inline src/position-observer.js]: Need to call `fire()` in `onLayoutMeasure()`. can't make it private

File: src/position-observer.js (hunk 2 above)
Comment (zhouyx) [inline src/position-observer.js]: I am OK with changing the name. However I am so bad at naming..... and honestly I like `PositionObserverApi` 😄  What would you suggest? `PositionApiEmitter`? (Note: it does listen for `send-positions` message to start sending position data)

File: src/position-observer.js (hunk 3 above)
Comment (zhouyx) [inline src/position-observer.js]: 👍  Sounds great!

File: src/position-observer.js (hunk 7 above)
Comment (zhouyx) [inline src/position-observer.js]: 👍  I changed to c.

File: src/position-observer.js (hunk 4 above)
Comment (alanorozco) [inline src/position-observer.js]: Cool. I think you should still call `this.fire` from the initial callback, though, to make sure that it's rate limited :)

File: test/integration/test-amp-ad-3p.js [hunk 10]
```diff
@@ -105,22 +104,14 @@ function createIframeWithApis(fixture) {
...
       lastIO = changes[changes.length - 1];
     });
-    iframe.contentWindow.context.observePosition(data => {
-      lastPosition = data.target;
-    });
     fixture.win.scrollTo(0, 1000);
     fixture.win.document.body.dispatchEvent(new Event('scroll'));
     return poll('wait for new IO entry', () => {
       return lastIO != null;
     });
```
Comment (alanorozco) [inline test/integration/test-amp-ad-3p.js]: Why did you remove these lines from the test?

File: test/functional/test-position-observer.js [hunk 11]
```diff
@@ -55,13 +54,12 @@ describes.realWin('PositionObserverApi', {
...
   });
 
-  it('should send one immediately wit correct value', () => {
-    const sendSpy = sandbox.spy(posObserver.subscriptionApi_, 'send');
-    posObserver.fire();
-    expect(sendSpy).to.be.calledOnce;
-    expect(sendSpy).to.be.calledWith('position', {
-      viewport: layoutRectLtwh(1, 2, 3, 4),
-      target: layoutRectLtwh(5, 6, 7, 8),
+  it('should get position with correct value', () => {
```
Comment (alanorozco) [inline test/functional/test-position-observer.js]: Why did you change this test?

File: src/position-observer.js [hunk 12]
```diff
@@ -47,44 +47,45 @@ export class PositionObserverApi {
...
   }
 
   /**
    * Clean all listeners
    */
   destroy() {
-    if (this.unlistenOnDestroy_) {
-      this.unlistenOnDestroy_();
-      this.unlistenOnDestroy_ = null;
+    for (let i = this.unlistenOnDestroy_.length - 1; i >= 0; i--) {
```
Comment (alanorozco) [inline src/position-observer.js]: Nit: to make the stack loop more expressive and concise you could do this instead:

```javascript
while (this.unlistenOnDestroy_.length) {
  this.unlistenOnDestroy_.pop().call();
}
```

File: src/position-observer.js (hunk 2 above)
Comment (alanorozco) [inline src/position-observer.js]: What about `PositionReporter`?

File: src/position-observer.js (hunk 4 above)
Comment (zhouyx) [inline src/position-observer.js]: Not calling `this.fire` to keep behavior consistency with inabox position observer 
https://github.com/ampproject/amphtml/blob/master/ads/inabox/position-observer.js#L70

File: src/position-observer.js (hunk 12 above)
Comment (zhouyx) [inline src/position-observer.js]: I like `while` better as well. Used `for` because I thought getting length is expensive. (I double checked, and it is only expensive for array-like object like NodeList. ). Will change.

File: test/integration/test-amp-ad-3p.js (hunk 10 above)
Comment (zhouyx) [inline test/integration/test-amp-ad-3p.js]: : ( Because it is failing, And I don't want to debug integration test ...

File: test/functional/test-position-observer.js (hunk 11 above)
Comment (zhouyx) [inline test/functional/test-position-observer.js]: Made this change , see [comments](https://github.com/ampproject/amphtml/pull/8708/files/ea827f64da0bb2a4bc6d19ea699ad3e1b63de474..40af82282974b5302a396a65a411a715a4d09bbc#r111459434). `fire()` will no longer covers the logic of `send one immediately without correct value`.

**Comment by zhouyx** [comment]
The PR is a bit mess. I need to keep consistency with existing inabox host positionObserver, then according to what we discussed, it seems we want to change inabox host positionObserver pretty soon. 😖 😢 @alanorozco I'll get back to your comments maybe tomorrow, after we figure out what the inabox host positionObserver should be.
